modules/renter/repair.go

```diff
@@ -1,365 +1,473 @@
-		r.repairChunks(f, handle, incChunks, pool)
-	}
+	// Log the error and retire the worker.
+	r.log.Debugln("Error while performing upload to", finishedUpload.workerID, "::", finishedUpload.err)
+	delete(rs.activeWorkers, finishedUpload.workerID)
+
+	// Indicate in the set of incomplete chunks that this piece was not
+	// completed.
+	rs.incompleteChunks[finishedUpload.chunkID].pieces[finishedUpload.pieceIndex] = struct{}{}
 }
 
-// repairChunks uploads missing chunks of f to new hosts.
-func (r *Renter) repairChunks(f *file, handle io.ReaderAt, chunks map[uint64][]uint64, pool *hostPool) {
-	for chunk, pieces := range chunks {
-		// Determine host set. We want one host for each missing piece, and no
-		// repeats of other hosts of this chunk.
-		hosts := pool.uniqueHosts(len(pieces), f.chunkHosts(chunk))
-		if len(hosts) == 0 {
-			r.log.Debugf("aborting repair of %v: host pool is empty", f.name)
-			return
+// threadedQueueRepairs is a goroutine that runs in the background and
+// continuously adds files to the repair loop, slow enough that it's not a
+// resource burden but fast enough that no file is ever at risk.
+//
+// NOTE: This loop is pretty naive in terms of work management. As the number
+// of files goes up, and as the number of chunks per file goes up, this will
+// become a performance bottleneck, and even inhibit repair progress.
+func (r *Renter) threadedQueueRepairs() {
+	for {
+		// Compress the set of files into a slice.
+		id := r.mu.RLock()
+		var files []*file
+		for _, file := range r.files {
+			files = append(files, file)
 		}
-		// upload to new hosts
-		err := f.repair(chunk, pieces, handle, hosts)
-		if err != nil {
-			if he, ok := err.(hostErrs); ok {
-				// if a specific host failed, remove it from the pool
-				for _, h := range he {
-					// only log non-graceful errors
-					if h.err != modules.ErrStopResponse {
-						r.log.Printf("failed to upload to host %v: %v", h.host, h.err)
-					}
-					pool.remove(h.host)
-				}
-			} else {
-				// any other type of error indicates a serious problem
-				r.log.Printf("aborting repair of %v: %v", f.name, err)
+		r.mu.RUnlock(id)
+
+		// Add one file every 30 seconds.
+		for _, file := range files {
+			// Send the file down the repair channel.
+			select {
+			case r.newRepairs <- file:
+			case <-r.tg.StopChan():
+				return
+			}
+
+			// Wait 30 seconds before going to the next file.
+			select {
+			case <-time.After(time.Second * 30):
+			case <-r.tg.StopChan():
 				return
 			}
 		}
 
-		// save the new contract
-		f.mu.RLock()
-		err = r.saveFile(f)
-		f.mu.RUnlock()
-		if err != nil {
-			// If saving failed for this chunk, it will probably fail for the
-			// next chunk as well. Better to try again on the next cycle.
-			r.log.Printf("failed to save repaired file %v: %v", f.name, err)
+		// Chill out for an extra 5 minutes before going through the files
+		// again.
+		select {
+		case <-time.After(time.Minute * 5):
+		case <-r.tg.StopChan():
```

Comment:
I wrote a nice [helper package](https://github.com/lukechampine/sleep/blob/master/sleep.go) for this sort of thing:
```go
interrupted := sleep.Lightly(5 * time.Minute, r.tg.StopChan())
if interrupted {
    return
}
```
(not that we should be importing it, but this pattern seems common enough that we may want to duplicate it)